### PR TITLE
Sharing all config volumes using virtiofs

### PR DESCRIPTION
**Pull request by germag**
**What this PR does / why we need it**:
Currently, using a virtiofs filesystem you can share ConfigMaps, Secrets and DownwardAPI volumes with the VM, but a ServiceAccount volume cannot be shared.

This PR allows sharing a ServiceAccount volume using a virtiofs filesystem and also mounts the config volumes in the source path defined for each of them. Additionally, it doesn't mount the volumes in the compute container if it's not necessary.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

**Special notes for your reviewer**:

Part of this PR was originally part of https://github.com/kubevirt/kubevirt/pull/9254

**Release note**:
-->
```release-note
NONE
```

### CODE REVIEW

**Comment by germag** [comment]
Rebase

**Comment by germag** [comment]
v2:
- Fix rendervolumes test
- Add sig tag to functional tests

**Comment by ormergi** [comment]
Hey, just to understand, what is a ServiceAccount volume?

**Comment by germag** [comment]
> Hey, just to understand, what is a ServiceAccount volume?

Hi, basically, is the way to access the tokens to authenticate to the API service (see https://kubernetes.io/docs/tasks/configure-pod-container/configure-service-account/)

These tokens could expire by time or become invalid if you migrate the VM to another node. The current method of using an iso image don't propagate the changes to the VM, so if these tokens expire or the VM is migrated the tokens inside the iso image becomes invalid.

Using virtiofs we can solve the propagation problem. Although virtiofsd doesn't yet support migration, we are working on it (with this use case in mind). We have a working prototype and a proposal for the changes that we need in qemu and the vhost-user protocol.

**Comment by alicefr** [comment]
/lgtm

Thanks @germag, looks good to me!

File: pkg/virt-controller/services/virtiofs.go [hunk 1]
```diff
@@ -84,34 +87,65 @@ func securityContextVirtioFS() *k8sv1.SecurityContext {
...
 }
 
-func virtioFSMountPoint(volumeName string) string {
-	return fmt.Sprintf("/%s", volumeName)
+func isConfig(volume *v1.Volume) bool {
+	return volume.ConfigMap != nil || volume.Secret != nil ||
+		volume.ServiceAccount != nil || volume.DownwardAPI != nil
+}
+
+func isAutomount(volume *v1.Volume) bool {
```
Comment (xpivarc) [inline pkg/virt-controller/services/virtiofs.go:95]: super nit: `isAutoMount`

File: tests/virtiofs/datavolume.go [hunk 2]
```diff
@@ -0,0 +1,289 @@
...
+	"kubevirt.io/kubevirt/tests/libstorage"
+	"kubevirt.io/kubevirt/tests/libwait"
+	"kubevirt.io/kubevirt/tests/testsuite"
+)
+
+const (
+	checkingVMInstanceConsoleOut = "Checking that the VirtualMachineInstance console has expected output"
+)
+
+var _ = Describe("[sig-storage] virtiofs", decorators.SigStorage, func() {
```
Comment (xpivarc) [inline tests/virtiofs/datavolume.go:62]: super nit: you can do `storage.SigDescribe` or just don't move this to your own package but rather your own file.

File: tests/virtiofs/config.go [hunk 3]
```diff
@@ -0,0 +1,272 @@
...
+		BeforeEach(func() {
+			// We use the Secret name as mount `tag` for qemu, but the `tag` property must be 36 bytes or less
+			secretName = "secret-" + uuid.NewRandom().String()[:6]
+			secretPath = config.GetSecretSourcePath(secretName)
+
+			data := map[string]string{
+				"user":     "admin",
+				"password": "redhat",
+			}
+			tests.CreateSecret(secretName, testsuite.NamespacePrivileged, data)
```
Comment (xpivarc) [inline tests/virtiofs/config.go:129]: Will this be cleaned?

File: tests/virtiofs/config.go [hunk 4]
```diff
@@ -0,0 +1,272 @@
...
+	"kubevirt.io/kubevirt/tests/exec"
+	"kubevirt.io/kubevirt/tests/testsuite"
+
+	"kubevirt.io/kubevirt/pkg/config"
+	"kubevirt.io/kubevirt/tests"
+	"kubevirt.io/kubevirt/tests/console"
+	"kubevirt.io/kubevirt/tests/libvmi"
+)
+
+var _ = Describe("[sig-compute] vitiofs config volumes", decorators.SigCompute, func() {
```
Comment (xpivarc) [inline tests/virtiofs/config.go:47]: I think we could use a table test in this case but let's have a look at it later.

**Review by xpivarc — APPROVED** [review]

**Comment by xpivarc** [comment]
@germag my only concern is https://github.com/kubevirt/kubevirt/pull/9432#discussion_r1158199913 everything else looks on the spot!

File: tests/virtiofs/config.go (hunk 3 above)
Comment (jcanocan) [reply tests/virtiofs/config.go:129]: As far as I was able to test, both the secret and the namespace are cleaned up. Those are the steps followed for testing this:

1. Focused the `With a single Secret volume` functest.
2. Targeted `make functest`
3. Ran `kubectl get namespaces. The command does not output `kubevirt-test-privileged` namespace.
4. Ran `kubectl get secrets --all-namespaces`. The command does not output any `secret-[random]` secret.

Therefore, we are convinced that both the secret and namespace are cleaned up after the test is executed.

File: tests/virtiofs/datavolume.go (hunk 2 above)
Comment (jcanocan) [reply tests/virtiofs/datavolume.go:62]: Sorry but we are not completely sure to understand your concern. Could you please elaborate it a little bit? Many thanks.

File: tests/virtiofs/config.go (hunk 3 above)
Comment (xpivarc) [reply tests/virtiofs/config.go:129]: Right they are namespaced but you could still get collision. Nevertheless, we can look at this later.

**Pull request merged**